core/src/main/java/org/kiwix/kiwixmobile/core/page/bookmark/BookmarksFragment.kt

```diff
@@ -19,8 +20,9 @@ class BookmarksActivity : PageActivity() {
   override val noItemsString: String by lazy { getString(R.string.no_bookmarks) }
   override val switchString: String by lazy { getString(R.string.bookmarks_from_current_book) }
   override val switchIsChecked: Boolean by lazy { sharedPreferenceUtil.showBookmarksAllBooks }
-  override fun injection(coreComponent: CoreComponent) {
-    activityComponent.inject(this)
+
+  override fun inject(baseActivity: BaseActivity) {
+    baseActivity.coreActivityComponent.inject(this)
```

Comment:
cached

Reply:
I tried using the cached component but the viewmodel crashes if I do that? Maybe it is not injected? Gimme a sec and I'll post the log.

Reply:
```kotlin
java.lang.IllegalArgumentException: unknown model class class org.kiwix.kiwixmobile.core.page.history.viewmodel.HistoryViewModel
        at org.kiwix.kiwixmobile.core.ViewModelFactory.create(ViewModelFactory.java:51)
        at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:187)
        at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:150)
        at org.kiwix.kiwixmobile.core.page.history.HistoryFragment$pageViewModel$2.invoke(HistoryFragment.kt:34)
        at org.kiwix.kiwixmobile.core.page.history.HistoryFragment$pageViewModel$2.invoke(HistoryFragment.kt:15)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at org.kiwix.kiwixmobile.core.page.history.HistoryFragment.getPageViewModel(Unknown Source:6)
        at org.kiwix.kiwixmobile.core.page.history.HistoryFragment.getPageViewModel(HistoryFragment.kt:15)
        at org.kiwix.kiwixmobile.core.page.PageFragment.onViewCreated(PageFragment.kt:132)
```

Reply:
That is pretty odd, I'll have a look

Reply:
This error is coming from ViewModelFactory, a pretty intimidating if small class. It operates using a Dagger feature known as `multibinding` where you want to inject a collection eg a Map and the map it gets is a class to a function that creates said class ie our ViewModels. We define the bindings in this map like so 

```
  @Binds
  @IntoMap
  @ViewModelKey(HistoryViewModel::class)
  abstract fun bindHistoryViewModel(historyViewModel: HistoryViewModel): ViewModel
```

In CoreViewmodleModule.

I don't know why this is complaining now, it might be worth sticking a debug line in ViewModelFactory.

What might be the issue is that cachedComponents are not actually `CoreActivityComponents` despite inheriting from them

Reply:
hmmm, ok I'll do some debugging.

Reply:
hmmm no the module is on CoreComponent and CoreComponent is used in the construction of Kiwix/CustomComponent but an annoying quirk of this cross module dependency sharing thing is that instead of it being implicit eg all dependencies in CoreComponent are usable in CoreActivityComponent because it is a Subcomponent but because Kiwix/Custom _depend on_ CoreComponent we have to expose the multition map binding in CoreComponent explicitly and that is why CoreComponent has all these methods

```
  fun zimReaderContainer(): ZimReaderContainer
  fun sharedPrefUtil(): SharedPreferenceUtil
  fun zimFileReaderFactory(): ZimFileReader.Factory
  fun storageObserver(): StorageObserver
  fun kiwixService(): KiwixService
  fun application(): Application
  fun bookUtils(): BookUtils
  fun dataSource(): DataSource
  fun fetchDownloadDao(): FetchDownloadDao
  fun newBookDao(): NewBookDao
  fun newLanguagesDao(): NewLanguagesDao
  fun recentSearchDao(): NewRecentSearchDao
  fun newBookmarksDao(): NewBookmarksDao
  fun connectivityManager(): ConnectivityManager
  fun context(): Context
  fun downloader(): Downloader
  fun bookDao(): BookDao
  fun bookmarksDao(): BookmarksDao
  fun notificationManager(): NotificationManager
```

That was a pretty big and very confusing info dump on Dagger, let me know if you want to hop on a call